src/Workspaces/Core/Portable/Workspace/Solution/CachedVersionTextSource.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A text source that keeps the text version cached.
+    /// </summary>
+    internal class CachedVersionTextSource : ValueSource<TextAndVersion>, ITextVersionable
```

Comment:
that assumes it can't be Func<SourceText> but it must be Func<Task<SouceText>>.

if we restrict it to just Func<SourceText> (in this particular case, we get SourceText by root.GetText(encoding)), it doesn't need to be Async.